lms/djangoapps/course_goals/views.py

```diff
@@ -102,3 +104,56 @@ def create(self, post_data):  # lint-amnesty, pylint: disable=arguments-differ
             'is_unsure': goal_key == GOAL_KEY_CHOICES.unsure,
         }
         return JsonResponse(data, content_type="application/json", status=(200 if goal else 201))  # lint-amnesty, pylint: disable=redundant-content-type-for-json-response
+
+
+class PopulateUserActivity(APIView):
+    """
+    API that allows external callers to populate the user activity table
+    """
+    authentication_classes = (JwtAuthentication, SessionAuthentication,)
+    permission_classes = (permissions.IsAuthenticated, IsStaffOrOwner,)
+
+    def post(self, request):
+        """
+        Handle the POST request
+
+        Populate the user activity table.
+        """
+        user_id = request.data.get('user_id')
+        course_key = request.data.get('course_key')
+
+        if not user_id or not course_key:
+            return Response(
+                'User id and course key are required',
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        try:
+            user_id = int(user_id)
+            user = User.objects.get(id=user_id)
+        except User.DoesNotExist:
+            return Response(
+                'Provided user id does not correspond to an existing user',
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        try:
+            course_key = CourseKey.from_string(course_key)
+        except InvalidKeyError:
+            return Response(
+                'Provided course key is not valid',
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        # Populate user activity for tracking progress towards a user's course goals
+        (activity_object_id, created) = UserActivity.populate_user_activity(user, course_key)
+        if activity_object_id:
+            data = {
+                'activity_object_id': activity_object_id
+            }
+            return JsonResponse(data, status=(201 if created else 200))
+        else:
+            return Response(
+                'User Activity was not populated',
+                status=status.HTTP_400_BAD_REQUEST,
```

Comment:
i added the waffle mostly as a temporary way for me to test the mobile endpoints from the mobile apps without breaking things in the meantime and was going to delete it once that was done.

I don't know if we need a waffle flag for this long term (or if you think we do, we can probably reuse the goals flag, since the goals feature won't work without this anyways)

Reply:
No I don't care about the fate of the waffle flag per se. I was just suggesting that it's an aggressive response to a `None` return from the method - that not recording an item of user activity was necessarily an error state. There could be valid reasons why you wouldn't (waffle is one example, maybe there are others).

But this was more of a nit / prompt to think about whether you want that to be an error state as you refactor things. At least document in the docstring that `None` means something went wrong - so people on both sides of the function know that's the contract. (Alternatively, if all those cases are actually bad inputs, you could have the method raise exceptions and catch it here - which would also let you send back or log a more specific-to-the-problem message here)

Reply:
ah, well currently if the method records activity or finds cached activity, it is returning the object, so it only returns none if the inputs were invalid (or with my modifications now if the user was masquerading)
but given that mobile app doesn't have masquerade i believe and if we're going to make this mobile specific, then it would still be true that this return value would correspond to some sort of invalid inputs
not sure if you think there is a response code that would be better suited to that?

Reply:
> not sure if you think there is a response code that would be better suited to that?

My beef was no so much with the specific response code, but with the design of the function's interface.

A return value of `None` is _fine_. But it doesn't distinguish between the benign "nothing needed doing, all good" (waffle case, masquerading, maybe similar future cases, etc) and the client error of "you passed in invalid arguments".

Some solutons:
- Merely document that `None` is actually an error state and should be treated as such. This method does not have (and apparently promises to never have) "nothing needed doing" benign states. That's the promise that I'm having a hard time with here - why make that? Why leave that gotcha for some future engineer? If you really do leave it like this, at least make heavily document that a `None` return is a client error and should be treated as such, so an engineer will hopefully know not to add a benign case in the future (though, if they want to, how would they do that? they'd need to refactor the interface along these lines below, yeah? or fake a non-`None` response)
- Trust that your code in this view that parses the user id and gets it from the db and parses the course-key would already catch the bad states. And just always return 2xx. Removing a need to distinguish between the error and benign cases.
- Treat caller-error cases in a more pythonic (but sometimes more annoying to code around) way by throwing an exception. Then this code could catch it and return a 400 (with the added benefit of a specific error string). This leaves `None` to signal the benign "nothing needed doing" state.

Reply:
I already had documentation around the None return in the method docstring, but its true that i'm already checking the input cases, masquerade isn't a feature in mobile and any other errors would throw an exception anyway, so I've modified the code with your approach from bullet #2